src/main/java/at/hannibal2/skyhanni/data/PetAPI.kt

```diff
@@ -35,31 +194,412 @@ object PetAPI {
+            name,
+            rarity,
+            null,
+            level,
+            0.0,
+            skin,
+        )
+    }
+
+    @HandleEvent
+    fun onCommandRegistration(event: CommandRegistrationEvent) {
+        event.register("shcalcpetxp") {
+            description = "Gets the pet xp from a given level and rarity."
+            category = CommandCategory.DEVELOPER_TEST
+            callback { testLevelToXP(it) }
+        }
+    }
+
+    private fun testLevelToXP(input: Array<String>) {
+        if (input.size >= 3) {
+            val level = input[0].toIntOrNull()
+            val rarity = LorenzRarity.getByName(input[1])
+            val petName = input.slice(2..<input.size).joinToString(" ")
+            if (level != null && rarity != null) {
+                val xp: Double = levelToXP(level, rarity, petName) ?: run {
+                    ChatUtils.userError("Invalid level or rarity.")
+                    return
+                }
+                ChatUtils.chat(xp.addSeparators())
+                return
+            }
+        }
+        ChatUtils.userError("Invalid Usage. /shcalcpetxp <level> <rarity> <pet>")
+    }
+
+    private fun levelToXP(level: Int, rarity: LorenzRarity, petName: String = ""): Double? {
+        val newPetName = petNameToFakeInternalName(petName)
+        val petObject = customXpLeveling?.getAsJsonObject(newPetName)
+
+        val rarityOffset = getRarityOffset(rarity, petObject?.getAsJsonObject("rarity_offset")) ?: return null
+        if (!isValidLevel(level, petObject)) return null
+
+        val xpList = xpLeveling + getCustomLeveling(petObject)
+
+        return xpList.slice(0 + rarityOffset..<level + rarityOffset - 1).sum().toDouble()
+    }
+
+    private fun isValidLevel(level: Int, petObject: JsonObject?): Boolean {
+        val maxLevel = petObject?.get("max_level")?.asInt ?: 100
+
+        return maxLevel >= level
+    }
+
+    private fun getCustomLeveling(petObject: JsonObject?): List<Int> {
+        return petObject?.getAsJsonArray("pet_levels")?.map { it.asInt } ?: listOf()
+    }
+
+    private fun getRarityOffset(rarity: LorenzRarity, petObject: JsonObject?): Int? {
+        return if (petObject == null) {
+            when (rarity) {
+                LorenzRarity.COMMON -> 0
+                LorenzRarity.UNCOMMON -> 6
+                LorenzRarity.RARE -> 11
+                LorenzRarity.EPIC -> 16
+                LorenzRarity.LEGENDARY -> 20
+                LorenzRarity.MYTHIC -> 20
+                else -> {
+                    ChatUtils.userError("Invalid Rarity \"${rarity.name}\"")
+                    null
+                }
+            }
+        } else {
+            petObject.entrySet().associate { (rarity, offset) ->
+                (LorenzRarity.getByName(rarity) ?: LorenzRarity.ULTIMATE) to offset.asInt
+            }[rarity]
+        }
+    }
+
+    private fun fireEvent(newPet: PetData?) {
+        val oldPet = pet
+        pet = newPet
```

Comment:
can move this line below the equals check

Reply:
the equals check should be made the first line in the function doing `if (newPet == pet) return`